aws/sandboxes/heka/output/kinesis.lua

```diff
@@ -0,0 +1,236 @@
+
+streamName              = "foobar"
+-- partitionField       = "Uuid"
+-- credentialProvider   = "INSTANCE"
+-- clientConfig         = nil
+-- roleArn              = nil
+
+-- Due to the way Kinesis is billed it makes senses to pack multiple items
+-- into a single message to reduce the PUT Payload Unit charges. Warning: if
+-- pack is greater than zero and and pack_partitions is true (multiple batches)
+-- the check pointing system cannot track them individualy i.e. in case of a
+-- failure items could be lost.
+pack_percentage = 0 -- fill percentage of the put request (0 (no packing) - <100).
+         -- The record won't be sent unless the current PUT Payload unit is this
+         -- percentage full or the record has reached its maximum size or the
+         -- timed out is triggered.
+pack_delimiter = nil    -- Used to add a record delimiter, if necessary
+pack_partitions = false -- create a batch for each partition
+
+-- Specify a module that will encode/convert the Heka message into its output representation.
+encoder_module = "encoders.heka.protobuf" -- default
+```
+--]]
+
+require "aws.kinesis"
+require "string"
+require "table"
+require "xxhash"
+
+local PUT_MAX   = 1024 * 1024
+local PUT_UNIT  = 25 * 1024
+
+local streamName            = read_config("streamName") or error"streamName must be set"
+local partitionField        = read_config("partitionField") or "Uuid"
+local credentialProvider    = read_config("credentialProvider") or "INSTANCE"
+local roleArn
+if credentialProvider == "ROLE" then
+    roleArn = read_config("roleArn") or error"roleArn must be set"
+end
+local clientConfig          = read_config("clientConfig") or {}
+assert(type(clientConfig) == "table", "invalid clientConfig type")
+
+local encoder_module = read_config("encoder_module") or "encoders.heka.protobuf"
+local encode = require(encoder_module).encode
+if not encode then
+    error(encoder_module .. " does not provide an encode function")
+end
+
+local pack_percentage = read_config("pack_percentage") or 0
+local pack_partitions = read_config("pack_partitions")
+assert(pack_percentage >= 0 and pack_percentage < 100, "0 <= pack_percentage < 100")
+pack_percentage = pack_percentage / 100 * PUT_UNIT
+local pack_delimiter  = read_config("pack_delimiter")
+local pack_delimiter_size = 0
+if pack_delimiter then pack_delimiter_size = #pack_delimiter end
+
+local producer = aws.kinesis.simple_producer(clientConfig, credentialProvider, roleArn)
+local partitions
+local partitions_size
+local function new_partition(key)
+    return {
+        cnt = 0,
+        size = 0,
+        timer = true,
+        msgs = {},
+        key = key
+    }
+end
+
+if pack_percentage > 0 then
+    local function get_shard_count()
+        local retry = 0
+        repeat
+            local osc, err = producer:open_shard_count(streamName)
+            if osc then return osc end
+            print("open_shard_count", err)
+            retry = retry + 1
```

Comment:
is this is a retryable error should we include a delay here?

Reply:
The throttling is in the underlying API call